core/src/main/java/io/questdb/griffin/engine/table/AsyncGroupByNotKeyedAtom.java

```diff
@@ -0,0 +1,157 @@
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreaker;
+import io.questdb.cairo.sql.StatefulAtom;
+import io.questdb.cairo.sql.SymbolTableSource;
+import io.questdb.griffin.PlanSink;
+import io.questdb.griffin.Plannable;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.PerWorkerLocks;
+import io.questdb.griffin.engine.functions.GroupByFunction;
+import io.questdb.griffin.engine.groupby.GroupByFunctionsUpdater;
+import io.questdb.griffin.engine.groupby.GroupByFunctionsUpdaterFactory;
+import io.questdb.std.BytecodeAssembler;
+import io.questdb.std.Misc;
+import io.questdb.std.ObjList;
+import io.questdb.std.Transient;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.Closeable;
+
+public class AsyncGroupByNotKeyedAtom implements StatefulAtom, Closeable, Plannable {
+
+    private final Function filter;
+    private final GroupByFunctionsUpdater functionUpdater;
+    private final ObjList<Function> perWorkerFilters;
+    private final PerWorkerLocks perWorkerLocks;
+    private final int valueCount;
+
+    public AsyncGroupByNotKeyedAtom(
+            @Transient @NotNull BytecodeAssembler asm,
+            @NotNull CairoConfiguration configuration,
+            @NotNull ObjList<GroupByFunction> groupByFunctions,
+            int valueCount,
+            @Nullable Function filter,
+            @Nullable ObjList<Function> perWorkerFilters,
+            int workerCount
+    ) {
+        assert perWorkerFilters == null || perWorkerFilters.size() == workerCount;
+        try {
+            this.filter = filter;
+            this.perWorkerFilters = perWorkerFilters;
+            this.valueCount = valueCount;
+            functionUpdater = GroupByFunctionsUpdaterFactory.getInstance(asm, groupByFunctions);
+            if (perWorkerFilters != null) {
+                perWorkerLocks = new PerWorkerLocks(configuration, workerCount);
+            } else {
+                perWorkerLocks = null;
+            }
+        } catch (Throwable e) {
+            close();
+            throw e;
+        }
+    }
+
+    public int acquire(int workerId, boolean owner, SqlExecutionCircuitBreaker circuitBreaker) {
+        if (perWorkerLocks == null) {
+            return -1;
+        }
+        if (workerId == -1 && owner) {
+            // Owner thread is free to use the original filter anytime.
+            return -1;
+        }
+        return perWorkerLocks.acquireSlot(workerId, circuitBreaker);
+    }
+
+    @Override
+    public void close() {
+        Misc.free(filter);
+        Misc.freeObjList(perWorkerFilters);
+    }
+
+    public Function getFilter(int slotId) {
+        if (slotId == -1) {
+            return filter;
+        }
+        if (perWorkerFilters != null) {
+            return perWorkerFilters.getQuick(slotId);
+        }
+        return null;
```

Comment:
Why this implementation is different from `AsyncGroupByAtom`: 

```Java
    public Function getFilter(int slotId) {
        if (slotId == -1 || perWorkerFilters == null) {
            return filter;
        }
        return perWorkerFilters.getQuick(slotId);
    }
```

and `return null;` line is not covered by tests

Reply:
That branch is never taken since `slotId` will be -1 when `perWorkerFilters` are null. It could be an `assert perWorkerFilters != null` assertion, but I find it a bit less future-proof than the code with no assertion. In case of `AsyncGroupByAtom` both `perWorkerFilters` and `perWorkerKeyFunctions` may be null independently, so `slotId != -1 && perWorkerFilters == true` can be true there.

Made code consistent in all atoms in b44fa4ee5d64ee3dd9d96b6560c79572c7bf074f